### PR TITLE
Refactor Scalar Operations

### PR DESCRIPTION
**Pull request by AlexandruGG**
This PR refactors Collection Operations which currently return a single scalar value inside a collection, returning the scalar directly.

* [x] Contains 
* [x] Every
* [x] Falsy
* [x] Has 
* [x] Match + static analysis check & type fixes
* [x] Nullsy
* [x] Truthy

Related to https://github.com/loophp/collection/discussions/107.

### CODE REVIEW

File: docs/pages/api.rst [hunk 1]
```diff
@@ -453,8 +453,7 @@ contains
 Check if the collection contains one or more values.
 
 .. warning:: The `values` parameter is variadic and will be evaluated as a logical ``OR``.
-             If you're looking for a logical ``AND``, you have to make separate calls to this method;
-             note the calls cannot be in succession because the collection will contain a boolean after the first call.
+             If you're looking for a logical ``AND``, you have to make separate calls to this method.
```
Comment (drupol) [inline docs/pages/api.rst:456]: I think we can get rid of this line, it doesn't make sense actually.

Comment (AlexandruGG) [reply docs/pages/api.rst]: you're right, I simplified this for the other operations that return booleans. It's quite obvious that separate calls should be made if you're trying to test multiple things as an `AND` if the operation returns a boolean 😄

File: src/Contract/Operation/Matchable.php [hunk 2]
```diff
@@ -20,9 +19,7 @@ interface Matchable
 {
     /**
      * @param callable(T, TKey, Iterator<TKey, T>): bool $callback
-     * @param null|callable(T, TKey, Iterator<TKey, T>): T $matcher
-     *
-     * @return Collection<int, bool>
+     * @param null|callable(T, TKey, Iterator<TKey, T>): bool $matcher
```
Comment (AlexandruGG) [inline src/Contract/Operation/Matchable.php:22]: this is now correct I believe? both the callback and the matcher should return booleans

**Comment by AlexandruGG** [comment]
I'm gonna give this one a final look a bit later before marking it ready for review

File: src/Contract/Operation/Matchable.php (hunk 2 above)
Comment (drupol) [reply src/Contract/Operation/Matchable.php:22]: Yes that's how it supposed to work.

**Comment by drupol** [comment]
> I'm gonna give this one a final look a bit later before marking it ready for review

Take your time mate, there is no hurry.

Do you think it would be possible to add some documentation that explains

* All the methods of the Collection returns a new collection
* Only the methods X,Y,Z returns something different
* Explaining the reasons of this breaking change.

I will also use that part in the changelog when doing the release later on.

**Comment by AlexandruGG** [comment]
> > I'm gonna give this one a final look a bit later before marking it ready for review
> 
> Take your time mate, there is no hurry.
> 
> Do you think it would be possible to add some documentation that explains
> 
> * All the methods of the Collection returns a new collection
> * Only the methods X,Y,Z returns something different
> * Explaining the reasons of this breaking change.
> 
> I will also use that part in the changelog when doing the release later on.

Yeah for sure! I was thinking to add a part in the documentation to explain about the operations but I forgot about it there for a moment 😄.

Regarding the last bullet point do you want that to be in the documentation or do you mean in the PR description?

**Comment by drupol** [comment]
Ok good 👍

I think we can do it in the documentation and I'll copy paste it for the changelog later.

**Comment by AlexandruGG** [comment]
> Ok good 👍
> 
> I think we can do it in the documentation and I'll copy paste it for the changelog later.

Sounds good. Btw for BC breaking changes like this should we have a new major version? I.e 5.0?

**Comment by drupol** [comment]
I don't know yet. We will see when it's done.

File: README.md [hunk 3]
```diff
@@ -76,7 +76,7 @@ This library has been inspired by:
     needed are standard PHP variables like `int`, `string`, `callable` or
     `iterator`.
     It allows users to use those operations individually, at their own will, to
-    build up something custom. Currently, more than [**80 operations**][28] are
+    build up something custom. Currently, more than [**100 operations**][28] are
```
Comment (AlexandruGG) [inline README.md:79]: did a new count and this is up over 100 now 🎉

**Comment by AlexandruGG** [comment]
> I don't know yet. We will see when it's done.

I think it depends if we want this project to follow [semver](https://semver.org/) or not. Because based on that any backwards-incompatible API changes should be done as part of a new **major** version.

Also if we do up the version when releasing this we should up the [docs](https://github.com/loophp/collection/blob/master/docs/conf.py#L60) version.

**Comment by AlexandruGG** [comment]
@drupol let me know what you think about the documentation update

File: README.md (hunk 3 above)
Comment (drupol) [reply README.md:79]: Excellent :-)

File: docs/pages/api.rst [hunk 4]
```diff
@@ -145,7 +145,32 @@ Another example
...
+or through the ``Collection`` object.
+
+.. literalinclude:: code/operations/background.php
+  :language: php
+
+When used separately, operations typically return a PHP `Generator`_ or an `Iterator`_.
+When used as a ``Collection`` method, operations fall into three main categories based on the return type:
+
+1. Operations that return a ``scalar`` value. Currently, this includes:
+   ``Contains``, ``Every``, ``Falsy``, ``Has``, ``Key``, ``Match`` (or ``MatchOne``), ``Nullsy``, ``Truthy``.
```
Comment (drupol) [inline docs/pages/api.rst:161]: I think the explanation is overall really great and I wouldn't change a thing, except for the classification points here.

What I propose is to list the operation types as such:

1. Operations which returns a boolean: countains, every, falsy, has... etc etc
2. Operations which returns a Collection of Collection: span, partition
3. Operation which returns an array: all
4. Operation which returns a Collection: all the rest.

(_be carefull _Collection::key()_ might return anything, not only a scalar, but anything_)

Also, what do you think about:

* Updating the signatures of all the operations in the API page and add their return type ?
or
* Updating all operations and add a note about the return type ?

(_about that page, it's very sad we cannot generate the documentation automatically from the classes themselves!_)

Let me know what you think.

**Review by drupol — CHANGES_REQUESTED** [review]

**Comment by drupol** [comment]
> > I don't know yet. We will see when it's done.
> 
> I think it depends if we want this project to follow [semver](https://semver.org/) or not. Because based on that any backwards-incompatible API changes should be done as part of a new **major** version.
> 
> Also if we do up the version when releasing this we should up the [docs](https://github.com/loophp/collection/blob/master/docs/conf.py#L60) version as well.

We should definitely follow semver... and therefore it's going to be version 5 indeed.

**Comment by AlexandruGG** [comment]
> > > I don't know yet. We will see when it's done.
> > 
> > 
> > I think it depends if we want this project to follow [semver](https://semver.org/) or not. Because based on that any backwards-incompatible API changes should be done as part of a new **major** version.
> > Also if we do up the version when releasing this we should up the [docs](https://github.com/loophp/collection/blob/master/docs/conf.py#L60) version as well.
> 
> We should definitely follow semver... and therefore it's going to be version 5 indeed.

Cool in that case I think we can do the `map` change as well to remove the [deprecation](https://github.com/loophp/collection/blob/master/src/Operation/Map.php#L48-L53) and only allow one callback, I'll do it in a future PR.

File: docs/pages/api.rst (hunk 4 above)
Comment (AlexandruGG) [reply docs/pages/api.rst]: @drupol thank you for the feedback, I've done some changes to the numbered list to incorporate your suggestions, with a couple tweaks. Let me know what you think of this new classification :).

Regarding updating the signatures of the operations in the docs -> it's something that I thought about today as well 😄 . I'll do it in a future PR for all of them

Comment (drupol) [reply docs/pages/api.rst]: Thanks all good now :)

Comment (AlexandruGG) [reply docs/pages/api.rst]: @drupol great, this is all ready on my side 😁

**Comment by drupol** [comment]
Thanks, a big one done :1st_place_medal:

**Pull request merged**